File: lib/brain.py
"""
brain.py — Persistent memory for the living agent.

The brain is a Markdown file that the agent reads before every interaction
and updates after. It stores: identity, ongoing tasks, user preferences,
recent conversation summaries, and learned patterns.

Usage:
    from lib.brain import Brain

    brain = Brain()                       # loads data/brain.md
    context = brain.get_context()         # markdown string for claude prompt
    brain.update("user_prefs", "Prefers Dutch in chat")
    brain.add_event("Built WhatsApp bot, merged PR #5")
"""

import os
import re
import datetime
from pathlib import Path

DEFAULT_BRAIN_PATH = Path(__file__).resolve().parent.parent / "data" / "brain.md"

# Maximum events to keep in recent history
MAX_EVENTS = 50


class Brain:
    """Persistent agent memory backed by a Markdown file."""

    def __init__(self, path: str | Path | None = None):
        self.path = Path(path) if path else DEFAULT_BRAIN_PATH
        self.path.parent.mkdir(parents=True, exist_ok=True)

        if not self.path.exists():
            self._init_brain()

        self._content = self.path.read_text()

    def _init_brain(self) -> None:
        """Create the brain file from the template."""
        template = Path(__file__).resolve().parent.parent / "data" / "brain.template.md"
        if template.exists():
            self.path.write_text(template.read_text())
        else:
            self.path.write_text(_DEFAULT_TEMPLATE)

    def reload(self) -> None:
        """Reload brain from disk (in case another process updated it)."""
        self._content = self.path.read_text()

    def save(self) -> None:
        """Write current brain state to disk."""
        self.path.write_text(self._content)

    def get_context(self) -> str:
        """Return the full brain content for injection into a claude prompt."""
        self.reload()
        return self._content

    def get_context_prompt(self) -> str:
        """Return a prompt-ready string that instructs Claude to use the brain."""
        brain_content = self.get_context()
        return (
            "You are a persistent AI agent. Below is your brain — your memory "
            "from previous sessions. Use it to maintain continuity. At the end "
            "of this interaction, you will update your brain with anything new "
            "you learned.\n\n"
            "<brain>\n"
            f"{brain_content}\n"
            "</brain>\n\n"
            "Important: Respond naturally as a continuous being. Reference past "
            "interactions when relevant. Remember user preferences."
        )

    def get_section(self, heading: str) -> str:
        """Extract content under a specific ## heading."""
        pattern = rf"## {re.escape(heading)}\n(.*?)(?=\n## |\Z)"
        match = re.search(pattern, self._content, re.DOTALL)
        return match.group(1).strip() if match else ""

    def update_section(self, heading: str, content: str) -> None:
        """Replace content under a specific ## heading."""
        pattern = rf"(## {re.escape(heading)}\n).*?(?=\n## |\Z)"
        replacement = rf"\g<1>{content}\n"
        new_content = re.sub(pattern, replacement, self._content, flags=re.DOTALL)
        if new_content == self._content:
            # Section didn't exist, append it
            self._content = self._content.rstrip() + f"\n\n## {heading}\n{content}\n"
        else:
            self._content = new_content
        self.save()

    def add_event(self, event: str) -> None:
        """Add a timestamped event to the Recent History section."""
        timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
        entry = f"- [{timestamp}] {event}"

        history = self.get_section("Recent History")
        lines = [l for l in history.split("\n") if l.strip()]
        lines.insert(0, entry)
        lines = lines[:MAX_EVENTS]

        self.update_section("Recent History", "\n".join(lines))

    def get_user_pref(self, key: str) -> str | None:
        """Get a specific user preference by key."""
        prefs = self.get_section("User Preferences")
        for line in prefs.split("\n"):
            if line.strip().startswith(f"- {key}:"):
                return line.split(":", 1)[1].strip()
        return None

    def set_user_pref(self, key: str, value: str) -> None:
        """Set a user preference (adds or updates)."""
        prefs = self.get_section("User Preferences")
        lines = [l for l in prefs.split("\n") if l.strip()]

        # Update existing or append
        updated = False
        for i, line in enumerate(lines):
            if line.strip().startswith(f"- {key}:"):
                lines[i] = f"- {key}: {value}"
                updated = True
                break
        if not updated:
            lines.append(f"- {key}: {value}")

        self.update_section("User Preferences", "\n".join(lines))


_DEFAULT_TEMPLATE = """# Agent Brain

## Identity
- Name: Atlas
- Role: Personal AI agent
- Platform: claude-vps-agent

## Active Tasks
No active tasks.

## User Preferences
- Language: auto-detect

## Learned Patterns
No patterns learned yet.

## Recent History
No events yet.
"""

File: src/lib/brain.ts
/**
 * brain.ts — Persistent memory for the living agent.
 *
 * The brain is a Markdown file that the agent reads before every interaction
 * and updates after. It stores: identity, ongoing tasks, user preferences,
 * recent conversation summaries, and learned patterns.
 */

import fs from "node:fs";
import path from "node:path";
import { withFileLockSync, atomicWrite } from "./filelock.js";

const PROJECT_ROOT = path.resolve(import.meta.dirname ?? path.dirname(new URL(import.meta.url).pathname), "..", "..");
const DEFAULT_BRAIN_PATH = path.join(PROJECT_ROOT, "data", "brain.md");

/** Maximum events to keep in recent history. */
const MAX_EVENTS = 50;

export class Brain {
  readonly path: string;
  private content: string;

  constructor(brainPath?: string) {
    this.path = brainPath ?? DEFAULT_BRAIN_PATH;
    const dir = path.dirname(this.path);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    if (!fs.existsSync(this.path)) {
      this.initBrain();
    }
    this.content = fs.readFileSync(this.path, "utf-8");
  }

  private initBrain(): void {
    const templatePath = path.join(PROJECT_ROOT, "data", "brain.template.md");
    if (fs.existsSync(templatePath)) {
      fs.writeFileSync(this.path, fs.readFileSync(templatePath, "utf-8"));
    } else {
      fs.writeFileSync(this.path, DEFAULT_TEMPLATE);
    }
  }

  /** Reload brain from disk (in case another process updated it). */
  reload(): void {
    this.content = fs.readFileSync(this.path, "utf-8");
  }

  /** Write current brain state to disk atomically. */
  save(): void {
    atomicWrite(this.path, this.content);
  }

  /** Return the full brain content for injection into a Claude prompt. */
  getContext(): string {
    this.reload();
    return this.content;
  }

  /** Return a prompt-ready string that instructs Claude to use the brain. */
  getContextPrompt(): string {
    const brainContent = this.getContext();
    return (
      "You are a persistent AI agent. Below is your brain — your memory " +
      "from previous sessions. Use it to maintain continuity. At the end " +
      "of this interaction, you will update your brain with anything new " +
      "you learned.\n\n" +
      "<brain>\n" +
      brainContent + "\n" +
      "</brain>\n\n" +
      "Important: Respond naturally as a continuous being. Reference past " +
      "interactions when relevant. Remember user preferences."
    );
  }

  /** Extract content under a specific ## heading. */
  getSection(heading: string): string {
    const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const pattern = new RegExp(`^## ${escaped}\n(.*?)(?=^## |$)`, "ms");
    const match = this.content.match(pattern);
    return match ? match[1].trim() : "";
  }

  /** Replace content under a specific ## heading (locked, atomic). */
  updateSection(heading: string, content: string): void {
    withFileLockSync(this.path, () => {
      this.reload();
      this.applySectionUpdate(heading, content);
      this.save();
    });
  }

  private applySectionUpdate(heading: string, content: string): void {
    const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const pattern = new RegExp(`(^## ${escaped}\n).*?(?=^## |$)`, "ms");
    const newContent = this.content.replace(pattern, `$1${content}\n`);

    if (newContent === this.content) {
      // Section didn't exist — check if heading is truly absent
      const headingPattern = new RegExp(`^## ${escaped}`, "m");
      if (!this.getSection(heading) && !headingPattern.test(this.content)) {
        this.content = this.content.trimEnd() + `\n\n## ${heading}\n${content}\n`;
      }
    } else {
      this.content = newContent;
    }
  }

  /** Add a timestamped event to the Recent History section. */
  addEvent(event: string): void {
    const now = new Date();
    const ts = [
      now.getFullYear(),
      String(now.getMonth() + 1).padStart(2, "0"),
      String(now.getDate()).padStart(2, "0"),
    ].join("-") + " " + [
      String(now.getHours()).padStart(2, "0"),
      String(now.getMinutes()).padStart(2, "0"),
    ].join(":");
    const entry = `- [${ts}] ${event}`;

    withFileLockSync(this.path, () => {
      this.reload();
      const history = this.getSection("Recent History");
      const lines = history.split("\n").filter((l) => l.startsWith("- "));
      lines.unshift(entry);
      const trimmed = lines.slice(0, MAX_EVENTS);
      this.applySectionUpdate("Recent History", trimmed.join("\n"));
      this.save();
    });
  }

  /** Get a specific user preference by key. */
  getUserPref(key: string): string | null {
    const prefs = this.getSection("User Preferences");
    for (const line of prefs.split("\n")) {
      if (line.trim().startsWith(`- ${key}:`)) {
        return line.split(":").slice(1).join(":").trim();
      }
    }
    return null;
  }

  /** Set a user preference (adds or updates). */
  setUserPref(key: string, value: string): void {
    const prefs = this.getSection("User Preferences");
    const lines = prefs.split("\n").filter((l) => l.trim());

    let updated = false;
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].trim().startsWith(`- ${key}:`)) {
        lines[i] = `- ${key}: ${value}`;
        updated = true;
        break;
      }
    }
    if (!updated) {
      lines.push(`- ${key}: ${value}`);
    }

    this.updateSection("User Preferences", lines.join("\n"));
  }
}

const DEFAULT_TEMPLATE = `# Agent Brain

## Identity
- Name: Atlas
- Role: Personal AI assistant
- Platform: Chat bot (Discord/Telegram)

## User Preferences
(Not yet configured — will be set during first conversation)

## Learned Patterns
No patterns learned yet.

## Recent History
No events yet.
`;
